streams-contrib/streams-processor-regex/src/main/java/org/apache/streams/regex/RegexUtils.java

```diff
@@ -74,10 +72,21 @@ private RegexUtils() {}
             //it if it ends up taking too long
             RegexRunnable regexRunnable = new RegexRunnable(pattern, content, capture);
 
-            threadPool.submit(regexRunnable);
-            threadPool.awaitTermination(REGEX_MATCH_THREAD_KEEP_ALIVE_DEFAULT_TIMEOUT, TimeUnit.MILLISECONDS);
+            ForkJoinTask<Map<String, List<Integer>>> task = threadPool.submit(regexRunnable);
 
-            return regexRunnable.getMatches();
+            Map<String, List<Integer>> matches = Maps.newHashMap();
+
+            try {
+                matches = task.get(REGEX_MATCH_THREAD_KEEP_ALIVE_DEFAULT_TIMEOUT, TimeUnit.MILLISECONDS);
+            } catch (Exception e) {
+                LOGGER.error("Timed out while trying to find matches for content: {} and pattern: {} Possible catastrophic backtracking!: {}", content, pattern, e);
+            } finally {
+                while(!task.isDone()) {
+                    task.cancel(true);
```

Comment:
Looking a bit closer at the docs, it seems like cancel() may not guarantee termination. Instead of looping, perhaps a single check, and then a cancel if necessary?